Fire events for hue remote buttons pressed

azogue opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->


## Proposed change
<!-- 
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->

* **Add hue remotes to hue integration**, supporting ZGPSwitch, ZLLSwitch and ZLLRotary switches as event generators for button presses, **following the `deCONZ` integration** which already handles these devices.
* Battery-powered remotes also get included in the device list with a battery sensor entity.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box! 
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example entry for `configuration.yaml`:
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR. Furthermore, for new integrations, it gives an impression of how
  the configuration would look like.
  Note: Remove this section if this PR does not have an example entry.
-->

```yaml
# Example configuration.yaml

```

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository] --> **TODO**

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

azogue committed:
Add remote platform to hue integration

supporting ZGPSwitch, ZLLSwitch and ZLLRotary switches.

* Ported from custom component Hue-remotes-HASS from @robmarkcole

azogue committed:
Add options flow for hue, to toggle handling of sensors and remotes

* Sensors are enabled by default, and remotes are disabled,
  to not generate any breaking change for existent users.
  Also, when linking a new bridge these defaults are used,
  so unless going explicitly to the Options menu,
  the old behavior is preserved.
* SensorManager stores the enabled platforms and ignores everything else.
* Bridge is created with flags for `add_sensors` and `add_remotes`,
  and uses them to forward entry setup to only the enabled platforms.
* Update listener removes disabled kinds of devices when options are changed,
  so device list is in sync with options, and disabled kinds disappear from HA,
  leaving the enable/disable entity option for individual devices.

azogue committed:
Fix hue bridge mock with new parameters

azogue commented:
#### Some screenshots from the new options panel and the generated hue device list on the background:

* Both remotes and sensors included:

<img width="1030" alt="remotes_and_sensors" src="https://user-images.githubusercontent.com/9748301/77628514-6faa9480-6f48-11ea-8910-2d239a8527d8.png">

* Both disabled:

<img width="1030" alt="both_disabled" src="https://user-images.githubusercontent.com/9748301/77628579-86e98200-6f48-11ea-9804-bf302f93a2fc.png">

* Only remotes:

<img width="1030" alt="only_remotes" src="https://user-images.githubusercontent.com/9748301/77628592-8c46cc80-6f48-11ea-9e4d-cfb38f27f7ce.png">

Kane610 commented:
Sorry, this will not fly. This is not what the remote platform is used for, which is to control other devices. If you want to add remote support they should be represented through the event system since they are only ever relevant in the moment. Though I don't know how much it will add when it is a polling solution, it would drive me mad probably :)

balloob commented on homeassistant/components/hue/__init__.py:
```diff
@@ -98,13 +106,50 @@
+                device_registry.async_remove_device(device_id)
+
+    # Update entry with new selected options and reload.
+    hass.config_entries.async_update_entry(entry, data={**entry.data, **entry.options})
```
I don't see any reason why we would store options in data. They are both already stored on the entry.

balloob commented on homeassistant/components/hue/__init__.py:
```diff
@@ -98,13 +106,50 @@
+
+    # Update entry with new selected options and reload.
+    hass.config_entries.async_update_entry(entry, data={**entry.data, **entry.options})
+    hass.async_create_task(hass.config_entries.async_reload(entry.entry_id))
```
Why would we reload this?

balloob commented on homeassistant/components/hue/bridge.py:
```diff
@@ -87,14 +97,26 @@ def host(self):
-            hass.config_entries.async_forward_entry_setup(
-                self.config_entry, "binary_sensor"
+
+        if self.add_sensors:
```
This PR should have been two PRs. One to add the options flow and one to add remotes.

I don't see the use of allowing to include sensors yes/no. If people really don't want to see it, they can disable the entities but they are not "in the way". 

What would be a use case to disable the sensors?

azogue replied:
> What would be a use case to disable the sensors?

As exposed in the long answer, I see 2 questions/decisions:
- Disable 1 entity --> forget about 1 item
- Disable 1 platform --> forget about 1 kind of items

In fact, the general idea was to _select_ the kind of devices you want to add/setup/handle

balloob commented on homeassistant/components/hue/remote.py:
```diff
@@ -0,0 +1,232 @@
+            attributes.update(
+                {
+                    "on": self.sensor.on,
+                    "reachable": self.sensor.reachable,
```
We don't store reachable.

balloob commented on homeassistant/components/hue/remote.py:
```diff
@@ -0,0 +1,232 @@
+        if hasattr(self.sensor, "battery"):
+            attributes.update(
+                {
+                    "on": self.sensor.on,
```
What is this value ? Why would we store it?

azogue replied:
No idea :=)

That comes from aiohue & CC. For motion sensors that is a main switch I think, but for remotes ¿?

robmarkcole replied:
The remotes were represented as sensors previously

balloob commented on homeassistant/components/hue/remote.py:
```diff
@@ -0,0 +1,232 @@
+    def device_state_attributes(self):
+        """Return the device state attributes."""
+        attributes = {
+            "model": self.sensor.type,
```
Model should be in device info. Not in attributes.

balloob commented on homeassistant/components/hue/remote.py:
```diff
@@ -0,0 +1,232 @@
+        attributes = {
+            "model": self.sensor.type,
+            "last_button_event": self.state or "No data",
+            "last_updated": self._parsed_sensor_last_updated(),
```
State in the state machine already has a last updated.

azogue replied:
Was my thought when I saw it, but this is another timestamp, also probably badly named. This is the real 'last_changed' stored in the bridge. It is not the same as our 'last_changed' or last_update, as they both refer to the state machine, not the sensor.

It is also used to force a state change event which would not appear otherwise when same button is pressed twice, I think (again, not sure as this is _inherit_ code)

balloob commented on homeassistant/components/hue/remote.py:
```diff
@@ -0,0 +1,232 @@
+    ].sensor_manager.async_register_component("remote", async_add_entities)
+
+
+class HueGenericRemote(GenericHueSensor, RemoteDevice):
```
This has nothing to do with a remote as remotes work in Home Assistant. Remote entities need to be able to control other things from within Home Assistant. Hue remotes cannot be controlled via the API, this should not be a remote entity.

balloob commented on homeassistant/components/hue/remote.py:
```diff
@@ -0,0 +1,232 @@
+        """Return the last button press of the remote."""
+        if self.sensor.modelid.startswith("RWL"):
+            return RWL_BUTTONS.get(self.sensor.state["buttonevent"])
+        return Z3_BUTTON.get(self.sensor.state["buttonevent"])
```
We never represent remotes that send events to Home Assistant as entities, they are fired as events. This should not be an entity.

balloob commented:
I don't see a use case for any of the remote entities nor do I see any use case for making sensors optional. Before you put any more time in this PR, let's first discuss the problems that you are trying to solve.

azogue commented:
> Sorry, this will not fly. This is not what the remote platform is used for, which is to control other devices. If you want to add remote support they should be represented through the event system since they are only ever relevant in the moment.

Hi @Kane610,

I know the approach for switches on ZHA is by triggering events, **as right now I'm migrating most of my hue remotes and motion sensors to deconz :)** (thanks to one ConBee II received just today).

I'm also not convinced about the 'remote' platform (I've been a long user of the custom integration for remotes when they were just 'sensors'), but this PR pretends to give support for those CC users, which now are **using entities with readable states** (like "1_click" vs 1001, etc).

> Though I don't know how much it will add when it is a polling solution, it would drive me mad probably :)

As pointed in the description, I agree that this polling system is not good for real-time automation, but:
- ZHA requires extra hardware and extra configuration, and right now is not so mainstream and user-friendly as managing devices on the Hue app, so for most users, that possibility is out
- Switch integration here in 'hue' enables a **_mixed_ integration which is not possible in ZHA** (as far as I know). 
For example, you can have the switch configured in the Hue app for simple presses (and handle simple light actions with them), and detect long presses in HA, to trigger scripts or more complex actions. For these, having some delay is not the end of the world, and, if we already have the motion sensors, why not these too?

cc @balloob 

> This has nothing to do with a remote as remotes work in Home Assistant. Remote entities need to be able to control other things from within Home Assistant. Hue remotes cannot be controlled via the API, this should not be a remote entity.

As said above, I agree with that. But, **How about making these as 'sensor' but maintain the toggle filters** (filtering 'sensor' kind on platform setup?)

> We never represent remotes that send events to Home Assistant as entities, they are fired as events. This should not be an entity.

An event is super for pushing integrations, but it doesn't store state nor attributes, so its usage is limited to trigger something. A 'sensor' entity covers more scenarios, IMO.

> I don't see a use case for any of the remote entities nor do I see any use case for making sensors optional. 

For _badly-named-remotes aka sensor switches_ I've exposed a use case (but I can imagine more I think). 
For the toggles to forget about sensors and/or remotes, well, I considered some assumptions (and now I'm discovering that some of them were wrong):
- If you have 'n' sensors, but you don't want them, and disable all of them, **the bridge is called anyway each 5s**, so there you have an extra request.
- Also, as our number of integrations increases, I personally don't like to see devices that I don't specifically want, but disabled devices appear anyway-
- Innocent of me, I thought the conbee worked with the same network that the Hue bridge, so you could see the devices in both --> today I discovered the truth :)
3rd point being wrong makes silly the 2nd point, as ported devices are not seen by the hue bridge anymore, but 1st point remains I think.

Also, the usage of the options flow was from my POV the easy way to maintain old behavior and be able to enable remotes from UI, and maybe open the way for migration to UI for other config options that right now are only accessible through manual config. In that context, toggling not only remotes but also sensors was a natural decision (I was going to move there the rest of config but I saw it out of context)

> Before you put any more time in this PR, let's first discuss the problems that you are trying to solve.

As you know, I've spent some time with the CC, and had some hope on the idea of a possible reduction in the polling scan_interval, but that dissipated completely when you pointed out the polling max limit in another PR, so I wasn't expecting any of that.

I also bought the conbee stick for my personal use :), and I'm now using it where I need no-delay action, but because of the mixed operation explained above, I will not move all gear to ZHA just now, and I thought others may be in similar situations, making this integration useful for them, and removing the need of the custom integration.

TBH, I pushed this today because I had it already done and it works, and IMO makes hue more _feature complete_. I didn't create tests or docs because I expected hard resistance :)

It is no problem for me to make these sensors and/or fire events, fix the option flow, make tests&docs, etc, as right now I'm _in confinement_ and with not much to do :)

But if there is no way to merge the hue dimmers with this approach, then close it.

Mariusthvdb commented:
> I don't see a use case for any of the remote entities nor do I see any use case for making sensors optional. Before you put any more time in this PR, let's first discuss the problems that you are trying to solve.

on the Hue remotes:
it would actually be nice to have a sensor (not remote) for the states/attributes of the hardware remotes, dimmer-switch and tap-switch mainly, and FOH added to that. For buttons pressed, and battery_level.
With these one could build further logic in HA automations (fully aware this would not be time critical because of the polling time), and battery warnings

on the Sensors being optional:
Yes please, this could be a big improvement over the current way of disabling and re-enabling the core sensors. It would allow for the Hue integration to only add the lights, or also add the sensors.
I happen to have a larger setup, in which the addition of the sensors takes a lot of the system, and need to disable the sensors, which I dont use. Having a simple toggle switch to help me do so would be a major improvement.

Hope you will consider.
thanks

balloob commented:
Let's add sensors with battery level for remotes and use events when we detect last pressed button has changed.

Remove the options. We are working hard in Home Assistant to harmonize all the things. That means that if you want to disable an entity, you go to the entity page and be able to disable any entity (even with multi-selecting entities) of any integration. We cannot have that some integrations just happen to have some extra options.

Mariusthvdb commented:
since you agree to add battery_level sensors, please let me ask if core Hue could simplify the current Hue sensors, which now all carry the same battery_level? Somewhat superfluous so it seems.

azogue commented:
> Let's add sensors with battery level for remotes and use events when we detect last pressed button has changed.

Ok, I'll do that (with a minor change: it is not the last press changing the thing that originates the event, but the change on `sensor.lastupdated` prop. If not, a repeated press goes unnoticed)

When you say 'sensors with battery level' you mean 1 sensor including batery_level attribute (and state showing last press) or something like the deconz integration, where only 1 battery entity is created for the device?
If it is the latter, then what should I do with remotes without battery level? (like the tap switch)

> Remove the options. 

Ok

> We are working hard in Home Assistant to harmonize all the things. That means that if you want to disable an entity, you go to the entity page and be able to disable any entity (even with multi-selecting entities) of any integration. We cannot have that some integrations just happen to have some extra options.

Wasn't aware of that, sorry. As explained in the comments, I thought disabling an entire platform could be interesting, in opposition to disable specific entities (which won't turn off the platform even when there are no enabled entities).

I took that path after reading the docs in https://developers.home-assistant.io/docs/entity_registry_disabled_by#integrations-offering-options-to-control-disabled_by, where it says "If this option is offered by integrations, you should not leverage the disabled_by property in the entity registry. Instead, if entities are disabled via a config options flow, remove them from the device and entity registry."



I'll address the proposed changes over the weekend.

Mariusthvdb commented:
> Ok, I'll do that (with a minor change: it is not the last press changing the thing that originates the event, but the change on sensor.lastupdated prop. If not, a repeated press goes unnoticed)

Hi @azogue,

thanks for taking this up, or could I say, for your determination ;-)

if you get to the sensors for last_updated response codes, would you please do so using the 'official' Hue api terminology, and not the 'fancified' hold_up variants...

`responsecodes = {"0":" click","1":" hold","2":" click release","3":" hold release"}`

Is what Hue is using for the RWL's, and it would be really nice to see that coming back , especially now a HA core integration seems so close!
thanks.

(in the above I've even left out the underscores, which is much better to the eye in the frontend

balloob commented:
@azogue dedicated battery sensors for devices that have battery. If a device does not have a battery, don't add a sensor. In that case do make sure you create the device via the 2nd example https://developers.home-assistant.io/docs/device_registry_index#defining-devices

Once we have a device for it, we can start adding device triggers too.

I will update the docs to clarify the section about disabling things.

azogue commented:
@Mariusthvdb, 

> if you get to the sensors for last_updated response codes, would you please do so using the 'official' Hue api terminology, and not the 'fancified' hold_up variants...

I'm going to use neither of those mappings. If we are going to homogeneize, we should maintain the same _schema_ for devices used over multiple integrations, IMO.  
The 'deconz_event's data contain 2 identifier keys and a _raw_ code, I think the new 'hue_event' should have the same contents (also to facilitate migrations).

The inclusion of _a new key in the event data_, with a string to map the button presses, I think it should be done in a separate PR, for all integrations firing these remote press events, so uniformity is maintained for known devices. What do you think?

@balloob, 

I'm not sure what is the best guide right now, if the ZHA or the deconz integration (I'm using the deconz right now)

balloob commented:
We don't want sensors to show last press, so just create an entity whose state is the current battery level and which has the battery device class.

azogue commented:
cc @balloob 

> We don't want sensors to show last press, so just create an entity whose state is the current battery level and which has the battery device class.

That was clear from your other message, I was talking about the event data, answering @Mariusthvdb, and proposing to include another extra field with some mapping for the remotes codes, to do it along with the multiple integrations that can handle these devices. Is that also a bad idea??

To you, I was just asking what is the best current integration to take it as guide, the ZHA or the deconz.

Mariusthvdb commented:
> We don't want sensors to show last press, 

@balloob Actually, 'we' do... Why would you believe that HA users don't need this?
Imho this would be a valuable addition to the core Hue sensors, and many users now forced to use the CC would be using this. There are many posts in the community on using this functionality, in very creative ways, related to Hue lights, but also many other actions triggered by their sensed presses on the tap-switch and dimmer-switches.

Please reconsider! Adding the simple sensor to Hue core for last press.

@azogue Maybe not the right place to talk about it here, but very related: Of course Paulus is right about the sensors for the dimmer-switch and tap-switch shouldn't be 'remotes'. In the CC this is causing confusion about their functionality and front-end representation also. 
Since it docent look like Paulus will agree on adding them to the core Hue sensor set, would you please discuss with @robmarkcole in rewriting the current fake 'remotes' back in to the proper sensor domain again. It used to be that way and was functioning very adequately.
thanks (and sorry for this longer post)

balloob commented:
@azogue every time we change data, we need to maintain it ourselves. If Hue has official definitions, we should use those, but I am assuming those are also returned from the API  ?

azogue commented:
> @azogue every time we change data, we need to maintain it ourselves. If Hue has official definitions, we should use those, but I am assuming those are also returned from the API ?

@balloob I think I got lost...

@Mariusthvdb said
> would you please do so using the 'official' Hue api terminology, and not the 'fancified' hold_up variants...

and I said
> I'm going to use neither of those mappings

The reason I said that is because:
* `aiohue` sensor definitions don't have any state mapping, official or otherwise. It works with the numeric raw codes, like '16' is the 2nd button of a Hue Tap switch, and '1001' is the 1st button release from a Hue dimmer switch.
* Other integrations that work with the same devices, like the deCONZ, are not mapping those codes, but instead are just using them, firing events like this one:

```json
{
    "event_type": "deconz_event",
    "data": {
        "id": "interruptor_cocina",
        "unique_id": "00:aa:bb:cc:dd:ee:ff:00",
        "event": 4002
    },
    "origin": "LOCAL",
    "time_fired": "2020-03-28T18:59:44.127882+00:00",
    "context": {
        "id": "d99a8b2d90a44811ad559da835f461a3",
        "parent_id": null,
        "user_id": null
    }
}
```

In that context, my intention was to also fire events from the hue integration, **maintaining the same info as the other integrations**. So the example above would be exactly the same, just changing `"event_type": "deconz_event"` for a new `"event_type": "hue_event"`.

And I suggested that adding **any** state mapping should be done in a separate PR, and involving not just hue but any integration handling these devices. And it could be done just by **adding** a new field in the event data. 

I didn't discuss anything about **what mapping** should be used, and yes, I agree with that, obviously:

> If Hue has official definitions, we should use those

And about this:
> but I am assuming those are also returned from the API ?

I think there is some information inside the `"capabilities"` field in the sensor raw data, but there is not a perfect mapping between raw code and pretty name, so I imagine that is the reason why those tags are not in use, and only the raw codes are passed on.

azogue commented:
[PARTIAL OFF-TOPIC]

@Mariusthvdb 

> Please reconsider! Adding the simple sensor to Hue core for last press.

I just did a quick custom integration ([azogue/eventsensor](https://github.com/azogue/eventsensor)) to solve this for those who want to see the last press as the state of some entity in HA (I'm included in that group too), so, if you want you can try it. It is already in the process to be [added to HACS](https://github.com/hacs/default/pull/301)

The entity is a sensor, and you can define any state mapping you want :)

I'm already using it for the remotes I moved to deCONZ

azogue commented:
BTW, 
> I just did a quick custom integration (azogue/eventsensor) to solve this

I chose the HACS path for that because I understood that kind of sensor is not desired in HA Core. If I'm wrong please correct me.

Mariusthvdb commented:
> [PARTIAL OFF-TOPIC]
> 
> @Mariusthvdb
> 
> > Please reconsider! Adding the simple sensor to Hue core for last press.
> 
> I just did a quick custom integration ([azogue/eventsensor](https://github.com/azogue/eventsensor)) to solve this for those who want to see the last press as the state of some entity in HA (I'm included in that group too), so, if you want you can try it. It is already in the process to be [added to HACS](https://github.com/hacs/default/pull/301)
> 
> The entity is a sensor, and you can define any state mapping you want :)
> 
> I'm already using it for the remotes I moved to deCONZ

can we take this elsewhere, not to spam the core HA/Hue discussion? (I like to try, but wouldn't know how to add the Hue switches that are now on the Hue Hub.)
In fact would rather have the original sensor.py modernized, and thus restore that faultless functionality...

azogue committed:
Revert changes in hue bridge mock

azogue committed:
Remove OptionsFlow and platform flags

azogue committed:
Extract `GenericHueDevice` from `GenericHueSensor`

to use it as base class for all hue devices, including those without any entity,
like remotes without battery.

azogue committed:
Add `HueBattery` sensor for battery powered remotes

and generate entities for TYPE_ZLL_ROTARY and TYPE_ZLL_SWITCH remotes.

azogue committed:
Remove remote platform

azogue committed:
Add HueEvent class to fire events for button presses

* Use `sensor.lastupdated` string to control state changes
* Event data includes:
  - "id", as pretty name of the remote
  - "unique_id" of the remote device
  - "event", with the raw code of the pressed button
    ('buttonevent' or 'rotaryevent' property)
  - "last_updated", with the bridge timestamp for the button press
* Register ZGP_SWITCH, ZLL_SWITCH, ZLL_ROTARY remotes

azogue committed:
fix removal

azogue committed:
Exclude W0611

balloob commented:
Oh you were talking about events. Yeah follow deconz.

azogue committed:
Extract GenericHueDevice to its own module

and solve import tree, also fixing lint in CI

balloob commented on homeassistant/components/hue/hue_event.py:
```diff
@@ -0,0 +1,101 @@
+        """Register callback that will be used for signals."""
+        super().__init__(sensor, name, bridge, primary_sensor)
+
+        self.event_id = slugify(self.sensor.name)
```
Why would we slugify it ?

azogue replied:
It is also done in deCONZ.
I suppose it is because the sensor name may contain non-standard chars (like 'ñ'), spaces, uppercase and so on, so in order to be used as the event 'id', it is slugified like any entity id, just for coherency.

balloob commented on homeassistant/components/hue/hue_event.py:
```diff
@@ -0,0 +1,101 @@
+        # Fire event
+        data = {
+            CONF_ID: self.event_id,
+            CONF_UNIQUE_ID: self.unique_id,
```
What's the difference between both ?

azogue replied:
* The first one (the _slugified_ one) is a user-given, friendly name, like "kitchen_remote". 
* The second one is the ugly mac/id of the device (`"00:aa:bb:cc:dd:ee:ff:00"`)

It is **like it's done in deCONZ**, and has its logic, as the pretty name is more useful to be used in automations and doesn't reveal any real identifier from the devices.

balloob commented on homeassistant/components/hue/hue_event.py:
```diff
@@ -0,0 +1,101 @@
+            CONF_LAST_UPDATED: self.sensor.lastupdated,
+        }
+        self.bridge.hass.bus.async_fire(CONF_HUE_EVENT, data)
+        _LOGGER.debug(
```
Why would we debug log this. Isn't this info also captured in the event? (which is also logged)

azogue replied:
Sorry, I put that while developing this (it is logged in core, which is not normally in debug level :)

I'll remove the line

balloob commented on homeassistant/components/hue/sensor_base.py:
```diff
@@ -64,9 +68,14 @@ def __init__(self, bridge):
         """Register async_add_entities methods for components."""
+        if platform not in self._enabled_platforms:
+            _LOGGER.debug("Aborting %s. It is not enabled", platform)
+            return
```
When can this happen ?

azogue replied:
Right now, never :)

I forget that when reverting the remote platform, I'll clean it

balloob reviewed (changes requested):
Oh forgot. Please add a test for this new functionality.

azogue commented:
> Oh forgot. Please add a test for this new functionality.

Sure. I will also need to add something about this on docs, right?

azogue commented on homeassistant/components/hue/sensor_device.py:
```diff
@@ -0,0 +1,53 @@
+
+    @property
+    def swupdatestate(self):
+        """Return detail of available software updates for this device."""
```
@balloob, this `swupdatestate` property was defined but unused in sensors. Should this be included in the `device_info` attributes?

balloob replied:
No. It's do indicate if a software update is available. I think that we check it for lights and notify user if there is one.

azogue replied:
> I think that we check it for lights and notify user if there is one.

That's right, it's the only usage, in https://github.com/home-assistant/core/blob/dd1608db0dfe0d140f9c5e92f57e97df5d87e46d/homeassistant/components/hue/light.py#L220

But there it is accessing the property from the `aiohue` object. 

This property is not called anywhere, that's why I was asking about

azogue committed:
Store registered events to do not repeat device reg

azogue committed:
Minor cleaning

azogue committed:
Add tests for hue_event and battery entities for hue remotes

balloob commented:
Great 🎉 

Now that there are events for remotes, we can add device triggers and so no one will ever see the events :) (that's same as with Deconz)

azogue commented:
> Now that there are events for remotes, we can add device triggers and so no one will ever see the events :)

It was going to be my next step :)

Should I do it before documenting this on docs, and then document the whole change, prioritizing the device triggers? (I didn't start with the docs for this yet :)

balloob commented:
We're tagging the beta tomorrow morning European time. So let's get this merged first. Not sure if device triggers make it before cut off?

balloob commented:
Went ahead and merged this, we can go do the docs once we have device triggers.

azogue commented:
> Not sure if device triggers make it before cut off?

Tomorrow I'll have more time than today, but I'll give it a try anyway :)

MartinHjelmare commented on tests/components/hue/test_sensor_base.py:
```diff
@@ -466,3 +501,121 @@ def mock_bridge(hass):
+    mock_bridge.mock_sensor_responses.append(new_sensor_response)
+
+    # Force updates to run again
+    await mock_bridge.sensor_manager.coordinator.async_refresh()
```
It's better to patch time and fire a time changed event to have the coordinator do a refresh. Then we're not relying on the details in our implementation in the tests.

azogue replied:
I just mirrored the way it is done in the other unit tests in this same file, where forced coordinator calls are being made like that :(

MartinHjelmare replied:
Yeah, my comment is for the future.

frenck commented:
Should this have been documented prior merging?

MartinHjelmare commented:
Custom events should normally be documented.

See https://github.com/home-assistant/core/pull/33277#issuecomment-606766186
